db_whitebox_test.go

```diff
@@ -112,6 +119,189 @@ func TestMethodPage(t *testing.T) {
 	}
 }
 
+func TestTx_Check_CorruptPage(t *testing.T) {
+	testCases := []struct {
+		name        string
+		bucketKey   string
+		expectPanic bool
+	}{
+		{
+			name:        "corrupt page by violating btree invariant",
+			bucketKey:   "testBucket",
+			expectPanic: false,
+		},
+		{
+			name:        "corrupt page by dumping random bytes",
+			bucketKey:   "testBucket",
+			expectPanic: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			fileName, err := prepareData(t)
+			require.NoError(t, err)
+
+			t.Logf("Creating db file '%v'.", fileName)
+			db, err := Open(fileName, 0666, &Options{PageSize: 4096})
+			require.NoError(t, err)
+			defer func() {
+				require.NoError(t, db.Close())
+			}()
+
+			uErr := db.Update(func(tx *Tx) error {
+				t.Logf("Creating bucket '%v'.", tc.bucketKey)
+				b, bErr := tx.CreateBucketIfNotExists([]byte(tc.bucketKey))
+				require.NoError(t, bErr)
+				t.Logf("Generating random data in bucket '%v'.", tc.bucketKey)
+				generateSampleDataInBucket(t, b, 3)
+				return nil
+			})
+			require.NoError(t, uErr)
+
+			t.Logf("Corrupting random leaf page in bucket '%v'.", tc.bucketKey)
+			victimPageId, validPageIds := corruptLeafPage(t, db, tc.expectPanic)
+
+			if tc.expectPanic {
+				defer func() {
+					r := recover()
+					require.NotNil(t, r)
+				}()
+			}
+
+			t.Log("Running consistency check.")
+			vErr := db.View(func(tx *Tx) error {
+				chkConfig := checkConfig{
+					kvStringer: HexKVStringer(),
+				}
+				ch := make(chan error)
+				defer close(ch)
+
+				var cErrs []error
+				go func() {
+					for cErr := range ch {
+						cErrs = append(cErrs, cErr)
+					}
+				}()
+				t.Log("Check corrupted page.")
+				chkConfig.pageId = uint(victimPageId)
+				tx.check(chkConfig, ch)
+				require.Equal(t, 1, len(cErrs))
+				require.ErrorContains(t, cErrs[0], fmt.Sprintf("leaf page(%d)", victimPageId))
+
+				t.Log("Check valid pages.")
+				cErrs = cErrs[:0]
+				for _, pgId := range validPageIds {
+					chkConfig.pageId = uint(pgId)
+					tx.check(chkConfig, ch)
+					require.Equal(t, 0, len(cErrs))
+				}
+				return nil
+			})
+			require.NoError(t, vErr)
+		})
+	}
+}
+
+// corruptLeafPage write an invalid leafPageElement into the victim page.
+func corruptLeafPage(t testing.TB, db *DB, expectPanic bool) (victimPageId common.Pgid, validPageIds []common.Pgid) {
+	t.Helper()
+
+	victimPageId, validPageIds = findVictimPageId(t, db)
+
+	victimPage, victimBuf, err := guts_cli.ReadPage(db.Path(), uint64(victimPageId))
+	require.NoError(t, err)
+	require.True(t, victimPage.IsLeafPage())
+	require.True(t, victimPage.Count() > 0)
+
+	// Dumping random bytes in victim page for corruption.
+	copy(victimBuf[32:], generateCorruptionBytes(t, expectPanic))
```

Comment:
@ahrtr @fuweid since we now check a page independently, I had to shift the corruption bytes one leafPageElement ahead

the reason is that the invalid leafPageElement I insert, needs the parent page to violate the ancestor key. Since each page is being checked independently, i shifted it one element ahead to cause corruption 👍🏽